system-addon/lib/SectionsManager.jsm

```diff
@@ -151,21 +151,37 @@ const SectionsManager = {
     }
   },
 
+  /**
+   * Save Top Story metadata to places db.
+   */
   updateBookmarkMetadata({url}) {
-    this.sections.forEach(section => {
-      if (section.rows) {
-        section.rows.forEach(card => {
-          if (card.url === url && card.description && card.title && card.image) {
-            PlacesUtils.history.update({
-              url: card.url,
-              title: card.title,
-              description: card.description,
-              previewImageURL: card.image
-            });
-          }
-        });
-      }
-    });
+    const {rows} = this.sections.get("topstories");
+    if (rows) {
+      rows.forEach(card => {
+        if (card.url === url && card.description && card.title && card.image) {
+          PlacesUtils.history.update({
+            url: card.url,
+            title: card.title,
+            description: card.description,
+            previewImageURL: card.image
+          });
+        }
+      });
+    }
+  },
+
+  /**
+   * When we bookmark a TopStory we need to add a visit to that URL
+   * to ensure it gets picked up in Highlights.
+   *
+   * @param {string} site.url
+   * @param {string} site.title
+   * @param {string} site.type Required to determine if it is a Top Story.
+   */
+  addBookmarkVisit({url, title, type}) {
+    if (type && (type === "trending" || type === "now")) {
```

Comment:
This seems like a pretty fragile way to determine if the thing being bookmarked is a story. We happen to only use those types for stories right now, but it doesn't seem unreasonable that Highlights or other sections might reuse those types.

On the other hand, if some other section say through webextension, allows a card to be bookmarked (are all cards bookmarkable?) would we expect those items to show up in Highlights if unvisited?

So I guess either the specificity of this behavior is too coupled with types when it should be checking for top-stories-sourced-cards or that it should be more general. If it should only be for stories, this action handler should probably be in `TopStoriesFeed` or if more general, in `PlacesFeed`.

Reply:
@Mardak would the correct behavior be to move this to PlacesFeed and add a history visit if we can't find any? Do you think that might be to expensive?

Reply:
Probably not worth it to check if visited first, but probably good to note it in a comment.